packages/core/src/new-api/internal/deployer.ts

```diff
@@ -62,11 +74,28 @@ export class Deployer {
       this._deploymentLoader.journal
     );
 
+    const contracts = getFuturesFromModule(module).filter(isContractFuture);
+    const contractStates = contracts.map(
+      (contract) => previousStateMap[contract.id]
+    );
+
+    // realistically this should be impossible to fail.
+    // just need it here for the type inference
+    assertIgnitionInvariant(
+      isContractExecutionStateArray(contractStates),
+      "Invalid state map"
+    );
+
+    const moduleArtifactMap = await this._loadModuleArtifactMap(contracts);
```

Comment:
Add a comment explaining that this was done like this because the reconciler is sync. If it were async we could pass it the loaders and it could load them JIT.